xwiki-rendering-syntaxes/xwiki-rendering-syntax-wikimodel/src/main/java/org/xwiki/rendering/internal/parser/wikimodel/AbstractWikiModelParser.java

```diff
@@ -131,11 +167,32 @@ public void parse(Reader source, Listener listener, IdGenerator idGenerator) thr
     {
         IWikiParser parser = createWikiModelParser();
         try {
-            parser.parse(source, createXWikiGeneratorListener(listener, idGenerator));
+            parser.parse(source, createXWikiGeneratorListener(buildListener(listener), idGenerator));
         } catch (Exception | StackOverflowError e) {
             // Stack overflow errors are caught in addition to exceptions because they can be thrown by javacc based
             // implementations in case of too deeply nested contents (e.g., too many nested groups).   
             throw new ParseException("Failed to parse input source", e);
         }
     }
+
+    private Listener buildListener(Listener listener)
+    {
+        // TODO: to be documented...
+        ListenerChain chain = new ListenerChain();
+        ChainingWrappingListener wrappedListener = new ChainingWrappingListener();
+        wrappedListener.setWrappedListener(listener);
+        StartChainingListener startChainingListener = new StartChainingListener();
+        startChainingListener.setListenerChain(chain);
+        chain.addListener(startChainingListener);
+        String roleHint = this.descriptor.getRoleHint();
+        Syntax syntax = this.syntaxRegistry.getSyntax(roleHint).orElse(null);
+        if (syntax == null) {
```

Comment:
It's a bit surprising to not use Optional#isEmpty since that was the kind of the point of returning an Optional in this API :)

Reply:
+1, though this is now replaced by the use of `resolveSyntax` which fallback to parsing if no matching syntax is registered, and throws a `ParseException` error if everything else failed.